handler/oauth2/flow_authorize_jwt_token.go

```diff
@@ -0,0 +1,330 @@
+		if err == nil {
+			return &key, nil
+		}
+	}
+
+	return nil, errorsx.WithStack(fosite.ErrInvalidGrant.WithHint(keyNotFoundMsg))
+}
+
+func (c *AuthorizeJwtGrantHandler) validateToken(ctx context.Context, token *jwt.JSONWebToken, key *jose.JSONWebKey) error {
+	claims := jwt.Claims{}
+	if err := token.Claims(key, &claims); err != nil {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.
+			WithHint("Unable to verify the integrity of the 'assertion' value.").
+			WithWrap(err).WithDebug(err.Error()),
+		)
+	}
+
+	if len(claims.Audience) == 0 {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.
+			WithHint("The JWT in \"assertion\" request parameter MUST contain an \"aud\" (audience) claim."),
+		)
+	}
+
+	if !claims.Audience.Contains(c.TokenURL) {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.
+			WithHintf(
+				"The JWT in \"assertion\" request parameter MUST contain an \"aud\" (audience) claim containing a value \"%s\" that identifies the authorization server as an intended audience.",
+				c.TokenURL,
+			),
+		)
+	}
+
+	if claims.Expiry == nil {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.
+			WithHint("The JWT in \"assertion\" request parameter MUST contain an \"exp\" (expiration time) claim."),
+		)
+	}
+
+	if claims.Expiry.Time().Before(time.Now()) {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.
+			WithHint("The JWT in \"assertion\" request parameter expired."),
+		)
+	}
+
+	if claims.NotBefore != nil && !claims.NotBefore.Time().Before(time.Now()) {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.
+			WithHintf(
+				"The JWT in \"assertion\" request parameter contains an \"nbf\" (not before) claim, that identifies the time '%s' before which the token MUST NOT be accepted.",
+				claims.NotBefore.Time().Format(time.RFC3339),
+			),
+		)
+	}
+
+	if !c.JWTIssuedDateOptional && claims.IssuedAt == nil {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.
+			WithHint("The JWT in \"assertion\" request parameter MUST contain an \"iat\" (issued at) claim."),
+		)
+	}
+
+	if claims.IssuedAt != nil && time.Now().Sub(claims.IssuedAt.Time()).Nanoseconds() > c.JWTMaxDuration.Nanoseconds() {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.
+			WithHintf(
+				"The JWT in \"assertion\" request parameter contains an \"iat\" (issued at) claim with value \"%s\" that is unreasonably far in the past.",
+				claims.IssuedAt.Time().Format(time.RFC3339),
+			),
+		)
+	}
+
+	var issuedDate time.Time
+	if claims.IssuedAt != nil {
+		issuedDate = claims.IssuedAt.Time()
+	} else {
+		issuedDate = time.Now()
+	}
+	if claims.Expiry.Time().Sub(issuedDate).Nanoseconds() > c.JWTMaxDuration.Nanoseconds() {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.
+			WithHintf(
+				"The JWT in \"assertion\" request parameter contains an \"exp\" (expiration time) claim with value \"%s\" that is unreasonably far in the future.",
+				claims.Expiry.Time().Format(time.RFC3339),
+			),
+		)
+	}
+
+	if !c.JWTIDOptional && claims.ID == "" {
```

Comment:
Can the JTI really be optional? How can we prevent replay attacks if it was optional?

Reply:
RFC tells that it is optional, that's why i made it so

Reply:
You mean the JWT spec, or the JWT-Bearer Auth Grant spec?

Reply:
I checked it now and both of them says it is optional